concepts/functions/about.md

```diff
@@ -1,3 +1,302 @@
+```python
+>>> add(2, 3)
+5
+
+# Function can be called multiple times, with different parameters
+>>> add(4, 3)
+7
+>>> add(5, 6)
+11
+
+# Passing incorrect number of parameters will raise `TypeError`
+>>> add(2)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: add() missing 1 required positional argument: 'y'
+
+>>> add(2, 3, 4)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: add() takes 2 positional arguments but 3 were given
+```
+
+## Return Value
+
+The return value is a value that is returned to the caller of the function. Return value can be of any data type. It can be used by caller of the function to perform further operations. If the function does not explicitly return a value, the value `None` is returned.
+
+Let's define a function `add`:
+```python
+def add(x, y):
+    return x + y
+```
+
+We'll store the return value in a variable and print it:
+```python
+>>> result = add(2, 3)
+>>> print(result)
+5
+# Type of result is `int`
+>>> type(result)
+<class 'int'>
+
+# We can also perform operations on the return value
+>>> result * 2
+10
+
+# Function without return value returns `None`
+>>> def log(message):
+    print(message)
+
+# Hello is printed because of print(message), but return value is `None`
+>>> return_value = log("Hello")
+Hello
+>>> return_value
+None
+```
+
+Use of `return` immediately exits the function and returns the value to the caller.
+```python
+>>> def show(x, y):
+    print(x)
+    return x
+    print(y)
+
+# y never gets printed, because function exists after return statement
+>>> show(1, 2)
+1
+```
+
+## Modularity
+
+Complex programs can be broken down into smaller parts. Functions can be used to perform different tasks.
+
+Assume a program has to perform the following tasks:
+  * Calculate the area of a circle
+  * Calculate the area of a rectangle
+  * Calculate the area of a triangle
+
+We can break down the program into smaller parts.
+
+```python
+def circle_area(r):
+    return 3.14 * r * r
+
+def rectangle_area(length, breadth):
+    return length * breadth
+
+def triangle_area(base, height):
+    return 0.5 * base * height
+```
+
+Now, we can call the functions in the order we want.
+
+```python
+>>> circle_area(2)
+12.56
+>>> rectangle_area(2, 3)
+6
+>>> triangle_area(2, 3)
+1.5
+```
+
+## Scope of Variables
+
+If variable is defined inside a function, then it will be only accessible inside the function. If we want to access the variable outside the function, we need to use the [`global`][global] keyword. [`nonlocal`][nonlocal] keyword is used to access the variable inside a nested function.
+
+```python
+def outer():
+    x = 10
+    print('Inside function:', x)
+```
+
+variable `x` is defined inside the outer function. It is limited to the scope of the outer function only. It will not alter the value of the variable outside the outer function.
+
+```python
+>>> x = 30
```

Comment:
You may want to move this line *before* the function definition so the reader knows that there's a global `x` before the function is shown.